cpp/src_prims/matrix/batched_matrix.hpp

```diff
@@ -0,0 +1,440 @@
+
+  //! Return allocator
+  std::shared_ptr<deviceAllocator> allocator() const { return m_allocator; }
+
+  //! return stream
+  cudaStream_t stream() const { return m_stream; }
+
+  //! Return shape
+  const std::pair<int, int>& shape() const { return m_shape; }
+
+  //! Return pointer array
+  double** data() const { return m_A_batches.get(); }
+
+  /* Return specific matrix pointer
+   * @param id Return the pointer to `id` matrix.
+   */
+  double* operator[](int id) const {
+    return &(m_A_dense.get()[id * m_shape.first * m_shape.second]);
+  }
+
+  //! Stack the matrix by columns creating a long vector
+  BatchedMatrix vec() const {
+    int m = m_shape.first;
+    int n = m_shape.second;
+    int r = m * n;
+    BatchedMatrix toVec(r, 1, m_num_batches, m_cublasHandle, m_allocator,
+                        m_stream);
+    cudaMemcpyAsync(toVec[0], this->operator[](0),
+                    m_num_batches * r * sizeof(double),
+                    cudaMemcpyDeviceToDevice, m_stream);
+    return toVec;
+  }
+
+  /* Create a matrix from a long vector.
+   * @param m Number of desired rows
+   * @param n Number of desired columns
+   */
+  BatchedMatrix mat(int m, int n) const {
+    int r = m_shape.first * m_shape.second;
+    BatchedMatrix toMat(m, n, m_num_batches, m_cublasHandle, m_allocator,
+                        m_stream);
+    cudaMemcpyAsync(toMat[0], this->operator[](0),
+                    m_num_batches * r * sizeof(double),
+                    cudaMemcpyDeviceToDevice, m_stream);
+
+    return toMat;
+  }
+
+  //! Visualize the first matrix.
+  void print(std::string name) const {
+    size_t len = m_shape.first * m_shape.second * m_num_batches;
+    std::vector<double> A(len);
+    updateHost(A.data(), m_A_dense.get(), len, m_stream);
+    std::cout << name << "=\n";
+    for (int i = 0; i < m_shape.first; i++) {
+      for (int j = 0; j < m_shape.second; j++) {
+        // column major
+        std::cout << std::setprecision(10) << A[j * m_shape.first + i] << ",";
+      }
+      std::cout << "\n";
+    }
+  }
+
+  //! Kernel to creates an Identity matrix
+  __global__ void identity_matrix_kernel(double** I, int r) {
+    int bid = blockIdx.x;
+    int tid = threadIdx.x;
+    for (int b = 0; b < r; b += blockDim.x) {
+      int idx = tid + b;
+      if (idx < r) {
+        I[bid][idx + r * idx] = 1;
+      }
+    }
+  }
+
+  /* Initialize a batched identity matrix.
+   * @param m Number of rows/columns of matrix
+   * @param num_batches Number of matrices in batch
+   * @param pool Memory pool
+   */
+  static BatchedMatrix Identity(int m, int num_batches, cublasHandle_t handle,
+                                std::shared_ptr<ML::deviceAllocator> allocator,
+                                cudaStream_t stream) {
+    BatchedMatrix I(m, m, num_batches, handle, allocator, stream, true);
+
+    identity_matrix_kernel<<<num_batches, std::min(1024, m), 0, stream>>>(
+      I.data(), m);
+
```

Comment:
it is good to put a CUDA_CHECK(cudaGetLastError()); here just to catch any issues with launching the kernel itself. (else becomes hard to pin-point the location of error, since these launches are host-asynchronous)